rig/routing_table/ordered_covering.py

```diff
@@ -0,0 +1,635 @@
+            # If the up-check removed any entries we need to re-perform the
+            # down-check; but we do not need to re-perform the up-check as the
+            # down check can only move the resultant merge nearer the top of
+            # the routing table.
+            merge = refine_downcheck(merge, aliases, min_goodness)
+
+    return merge
+
+
+def overlap(key_a, mask_a, key_b, mask_b):
+    return (key_a & mask_b) == (key_b & mask_a)
+
+
+def refine_upcheck(merge, min_goodness):
+    """Remove from the merge any entries which would be covered by entries
+    between their current position and the merge insertion position.
+
+    For example, the third entry of::
+
+        0011 -> N
+        0100 -> N
+        1000 -> N
+        X000 -> NE
+
+    Cannot be merged with the first two entries because that would move it
+    below the entry with the key-mask pair of ``X000``, which would cover it.
+
+    Returns
+    -------
+    :py:class:`~.Merge`
+        New merge with entries possibly removed.
+    bool
+        If the merge has been changed at all.
+    """
+    # Remove any entries which would be covered by entries above the merge
+    # position.
+    changed = False
+    for i in sorted(merge.entries, reverse=True):
+        # Get all the entries that are between the entry we're looking at the
+        # insertion index of the proposed merged index. If this entry would be
+        # covered up by any of them then we remove it from the merge.
+        entry = merge.routing_table[i]
+        key, mask = entry.key, entry.mask
+        if any(overlap(key, mask, other.key, other.mask) for other in
+               merge.routing_table[i+1:merge.insertion_index]):
+            # The entry would be partially or wholly covered by another entry,
+            # remove it from the merge and return a new merge.
+            merge = Merge(merge.routing_table, merge.entries - {i})
+            changed = True
+
+            # Check if the merge is sufficiently good
+            if merge.goodness <= min_goodness:
+                merge = Merge(merge.routing_table)  # Replace with empty merge
+                break
+
+    # Return the final merge
+    return merge, changed
+
+
+def refine_downcheck(merge, aliases, min_goodness):
+    """Prune the merge to avoid it covering up any entries which are below the
+    merge insertion position.
+
+    For example, in the (non-orthogonal) table::
+
+        00001 -> N S
+        00011 -> N S
+        00100 -> N S
+        00X00 -> N S
+        XX1XX -> 3 5
+
+    Merging the first four entries would generate the new key-mask ``00XXX``
+    which would be inserted above the entry with the key-mask ``XX1XX``.
+    However ``00XXX`` would stop the key ``00110`` from reaching its correct
+    route. To avoid this one could just abandon the merge entirely, but a
+    better solution is to attempt to reduce the merge such that it no longer
+    covers any entries below it.
```

Comment:
I'm very bad with terminology. Combined with the aliases being a verb and a noun (pl)...